### ISSUE TITLE
A few accessibility issues

### ISSUE THREAD

**Issue by rpkoller** [aatishb/covidtrends#123]
Aside the fact that a graph basically isn't accessible for blind people in general the general accessibility of the page could be improved never the less. 

- Three buttons don't have enough color contrast: 
![Bildschirmfoto 2020-04-22 um 16 55 26](https://user-images.githubusercontent.com/1665422/79997679-1b002800-84ba-11ea-9264-8491be7edb5f.png)
![Bildschirmfoto 2020-04-22 um 16 55 30](https://user-images.githubusercontent.com/1665422/79997687-1c315500-84ba-11ea-8fe6-1308f658bd44.png)
The color contrast in the header isn't possible to check due to the color gradient of the background
- the id `buttonwrap` is used twice in the `index.html` IDs should only be once. (By the way i would try to avoid IDs wherever it is possible) ;) 
- the nav element for pause and play should have an alt attribute with play/pause or if you place svg markup directly in the `index.html` an `aria-label` attribute in the svg markup. 
- it is recommended to provide labels for select boxes and fields. in the current state you have to guess what the select box is about based on the default value of the selectbox and on the placeholder text for the search field. also the timeline and the select box at the bottom of the page lack some sort of label. 
- aside that "search" is generic naming instead be more explicit and state something like "Search for country". and the attribute `type=search` might be also helpful for that particular input element
- then it would be also beneficial to utilize html5 elements like main, aside, footer, header and so forth. the `main` element is even recommended to markup the main content of the page. 
- the heading structure i wouldn't markup the date with a `h2`. The `h1` is ok and I would mark the `Trajectory of COVID-19....` headline as `h2`and in the sidebar maybe as `h2`as well or as `h3`

Just a few points to start with ;)

**Comment by rpkoller** [comment]
In regards of accessibility and Covid19 also an article: https://themarkup.org/2020/04/21/blind-users-struggle-with-state-coronavirus-websites

**Comment by mustpax** [comment]
Thank you @rpkoller! We agree that these are important issues to fix. Would you be open to submitting a PR to address any of these?

Please hold off on making any color scheme changes at this time, we are reevaluating this on our end.

**Comment by mustpax** [comment]
Also tracking similar accessibility issues detected by Google Chrome's Accessbility Audit tool under #129

**Comment by rpkoller** [comment]
@mustpax yep i would give an initial commit a spin. and don't worry about color changes those i wouldnt have touched anyway, that is in your field ;) especially if the background color of the button persists in that redish color then the according text color is tricky to find with enough of contrast. but when you are reevaluating colors then also consider change the color of the header to a solid color and remove the gradient. 

on a side note maybe even consider create a second page with documentation, some sort of guide, faq and or background informations. a clear separation of concerns and easier to digest for visitors. i mean that kind of graphs aren't easily accessible even for people without any temporary or permanent cognitive disabilities. those numbers and how to read them are hard to grasp anyway. ;) and some people prefer to read instead of watch a video. 

about the lighthouse test i am aware of it. it is based on axe core like accessibility insights by microsoft is (https://accessibilityinsights.io). but the latter pushes things a few steps further and provides some sort of guided manual testing protocol alongside as well. that is my usual go to ;)  

and one last thought. i have no idea if it would be accomplishable but at the moment more or less no graph on the net is really prepared for real blind users especially dynamically generated graphs like covidtrends. a recommended practice is if you have a svg or png of a graph alongside to the alt attribute also place a link right after the graph linking to an extensive description. so i was thinking about the following idea:
-generate such a descriptive text automatically based for each country selected. 
- describe todays current state with the number of overall infected persons and the number of infected persons last week and the number of deaths to date
- and if possible either state currently the growth is exponential or the growth is non exponential 
and slowing down... unsure if it is possible to calculate it in an either/or manner. 
there are many possibilities. but you could basically create snippets of text templates describing those different curves which are basically following all the same patterns. so if you could calculate and classify a curve with exponential growth or slowed down to non exponential growth and so on it would be possible. but that is just a future idea. ;) 

but i will start with the things i've described in my initial ticket first. ;) 

p.s. would it be ok to reduce the number of CSS IDs (change them to classes or rely even on element selectors)? and leave only those in which are used by vue.js? cuz normally the best practice is to avoid IDs due to their high specificity.

**Comment by aatishb** [comment]
Thanks @rpkoller for pointing out all these accessibility issues! Will be great to prioritize improving the site accessibility. Replying to some of your comments below.

> @mustpax yep i would give an initial commit a spin. and don't worry about color changes those i wouldnt have touched anyway, that is in your field ;) especially if the background color of the button persists in that redish color then the according text color is tricky to find with enough of contrast. but when you are reevaluating colors then also consider change the color of the header to a solid color and remove the gradient.

I think @mustpax did a contrast test to determine that the header contrast should be ok? (by checking the contrast between the font color and lightest color in the background gradient)

We can definitely reconsider the button background color if it becomes difficult to create a sufficiently high contrast ratio.

> on a side note maybe even consider create a second page with documentation, some sort of guide, faq and or background informations. a clear separation of concerns and easier to digest for visitors. i mean that kind of graphs aren't easily accessible even for people without any temporary or permanent cognitive disabilities. those numbers and how to read them are hard to grasp anyway. ;) and some people prefer to read instead of watch a video.

This sounds like a good idea for a separate issue. It will also fall under our clarity [milestone](https://github.com/aatishb/covidtrends/milestones). I have been meaning to eventually add more text-based explanatory information and possibly links to related resources.

> 
> about the lighthouse test i am aware of it. it is based on axe core like accessibility insights by microsoft is (https://accessibilityinsights.io). but the latter pushes things a few steps further and provides some sort of guided manual testing protocol alongside as well. that is my usual go to ;)

Thanks for sharing this resource!

> and one last thought. i have no idea if it would be accomplishable but at the moment more or less no graph on the net is really prepared for real blind users especially dynamically generated graphs like covidtrends. a recommended practice is if you have a svg or png of a graph alongside to the alt attribute also place a link right after the graph linking to an extensive description. so i was thinking about the following idea:
>    * generate such a descriptive text automatically based for each country selected.
> 
>    * describe todays current state with the number of overall infected persons and the number of infected persons last week and the number of deaths to date
> 
>    * and if possible either state currently the growth is exponential or the growth is non exponential
>       and slowing down... unsure if it is possible to calculate it in an either/or manner.
>       there are many possibilities. but you could basically create snippets of text templates describing those different curves which are basically following all the same patterns. so if you could calculate and classify a curve with exponential growth or slowed down to non exponential growth and so on it would be possible. but that is just a future idea. ;)
> 

I really like this idea and I'm interested in exploring how feasible this might be. Also may be worth creating a separate issue for it.

> but i will start with the things i've described in my initial ticket first. ;)
> 
> p.s. would it be ok to reduce the number of CSS IDs (change them to classes or rely even on element selectors)? and leave only those in which are used by vue.js? cuz normally the best practice is to avoid IDs due to their high specificity.

I'm fine with reducing the CSS ids and favoring classes wherever it makes sense.

Thanks @rpkoller! :)

**Comment by rpkoller** [comment]
> I think @mustpax did a contrast test to determine that the header contrast should be ok? (by checking the contrast between the font color and lightest color in the background gradient)
> 
> We can definitely reconsider the button background color if it becomes difficult to create a sufficiently high contrast ratio.

I was a bit premature with my suggestion. After you mentioned I've checked the font color as well as the read more color against the light and the dark color of the gradient. All had enough contrast. So it should be "ok". But never the less a gradient brings in possible distraction in contrast to a solid background color. For content things should be visually reduced as possible. But I will get a feedback on that over the weekend. 

> This sounds like a good idea for a separate issue. It will also fall under our clarity milestone. I have been meaning to eventually add more text-based explanatory information and possibly links to related resources.

You or me could open a new issue and layout the content and talking points there and then it could be discussed on. Especially there are ways to test if the finished copy is accessible and graspable. and so forth. ;) 

> I really like this idea and I'm interested in exploring how feasible this might be. Also may be worth creating a separate issue for it.

I'll open a new issue on the weekend and sum up and rephrase my initial thoughts from my post in here. But @aatishb is it basically possible to calculate easily and reliably if a curve overall or in a certain segment is in a state of exponential growth or not? 

And one question about the a11y related commits. Shall I create a new branch and work on the current state the master branch is in or should i work upon the reference line branch? Checked the changes for the reference line branch and they are only in `index.hml` and  `vue-definitions.js` and the changes in `index.html` consist only of adding the new checkbox markup. So there shouldn't be any conflicts if i would base the branch on the `master` one.

**Comment by aatishb** [comment]
> But never the less a gradient brings in possible distraction in contrast to a solid background color. For content things should be visually reduced as possible. But I will get a feedback on that over the weekend.

Thanks!

> I'll open a new issue on the weekend and sum up and rephrase my initial thoughts from my post in here. But @aatishb is it basically possible to calculate easily and reliably if a curve overall or in a certain segment is in a state of exponential growth or not?

In theory, yes, I think we could look at recent data points and infer whether the cases are growing exponentially, growing linearly, or falling. Of course there might be issues implementing this in practice. It might be safer to make a cruder statement, like compare 1 week ago to today and say whether the # of new cases are rising, falling, or roughly staying the same. e.g. the NYT does some summaries like this [here](https://www.nytimes.com/interactive/2020/04/23/upshot/five-ways-to-monitor-coronavirus-outbreak-us.html) (section 3 onwards).

> And one question about the a11y related commits. Shall I create a new branch and work on the current state the master branch is in or should i work upon the reference line branch? Checked the changes for the reference line branch and they are only in `index.hml` and `vue-definitions.js` and the changes in `index.html` consist only of adding the new checkbox markup. So there shouldn't be any conflicts if i would base the branch on the `master` one.

master branch is fine, since the reference line branch is still a work in progress. I'm also not worried about merge conflicts, we can resolve them if needed.

**Comment by rpkoller** [comment]
> In theory, yes, I think we could look at recent data points and infer whether the cases are growing exponentially, growing linearly, or falling. Of course there might be issues implementing this in practice. It might be safer to make a cruder statement, like compare 1 week ago to today and say whether the # of new cases are rising, falling, or roughly staying the same. e.g. the NYT does some summaries like this here (section 3 onwards).

I'll take a look at that article. Thanks! 

> master branch is fine, since the reference line branch is still a work in progress. I'm also not worried about merge conflicts, we can resolve them if needed.

ok i've started to dig into. one question. would it be ok to also clean up the indentation of the `index.html` as well as remove the empty lines?

**Comment by aatishb** [comment]
Maybe hold off on the html clean up for now? That'll keep the diff more focused here. I'm thinking once we're done merging a few more PRs we can do some cleanup.

**Comment by rpkoller** [comment]
yep you are right. is just a reflex to care about indentation and leave unnecessary empty lines out. but would overcomplicate things for the moment. but we should keep it in mind. 

by the way what is the purpose of the #title ID? there is no usage in `index.html` and in the `vue-definitions.js` file i am not entirely sure if needed and or used

**Comment by rpkoller** [comment]
@aatishb could you perhaps create a branch `accessibility` in this repo here. then i could place the pull request against the `accessibility` branch instead of the `master` branch. makes things easier afterwards with testing and having a look first. 

I've fixed all issues also listed in @mustpax report. except the color contrast that you two would have to check stylistically first which road you wanna go. aside that i thought i commit the first batch of changes so things go in smaller steps instead of one huge change. 

there are still a few fields for the future. keyboard navigation especially the tab order. unsure if it is possible to exclude the plotly ones. the we can also check outlines as well as responsive design in general. as well as link styling. still a few more topics to cover afterwards

**Comment by aatishb** [comment]
Thanks @rpkoller! I created the accessibility branch.

**Comment by rpkoller** [comment]
@aatishb thanks for the fast creation. Is cleaner that way and things stay a bit more sane ;)

**Comment by rpkoller** [comment]
@aatishb in regards of describing graphs and curves aside the idea i head with the text templates based on the gist of the curve you might also describe it by sonification: https://www.highcharts.com/docs/accessibility/sonification another take how to make a graph more accessible ;)

**Comment by rpkoller** [comment]
In regards of the markup with `main` `aside` I've started a question on Stackoverflow and got a few profound feedback. Will read up on those points a little further and the incorporate those suggested changes in a new pull request.

**Comment by rpkoller** [comment]
ok. it looks the `aside` is semantically incorrect. the cleaner way to go is to use the `main` tag and use the `section` element for the graph as well as for the sidebar with its controls and apply `wai-aria` attributes. tried it out locally already. 
the only additional thing to discuss, it would make sense to add a proper heading for the graph `section`. at the moment that "heading" is generated within the graphs canvas (" trajectory of world..."). adding an `h2` tag in there seems wrong if it would be possible at all. instead it would be more clean to simply create an `h2` tag right before the graph within the graphs `section`... would look that way (without any special styling or adjustment yet) : 

![Bildschirmfoto 2020-04-30 um 21 50 07](https://user-images.githubusercontent.com/1665422/80850932-ebba8b00-8c1e-11ea-9c36-816d068a23b5.png)

would be more consistent. only downside. vertically it takes more space because the headline is stacked on top of the Plotly toolbar which is stacked on top of the graph. :/ Maybe move the headline away from the center to the left on the same height as the toolbar if possible (not tried that yet)?

**Comment by rpkoller** [comment]
@aatishb wasn't starting to work on that part before i get any feedback. ;) any thoughts on that?

**Comment by aatishb** [comment]
Hi @rpkoller, sorry for the slow response. To be honest I'm a bit over my head here as I don't have any experience with this. I don't understand why main/section is preferred over aside, and also don't really know what wai-aria is. Is there something I could read to learn more about this? I am happy to follow your lead on the semantic HTML changes towards improving accessibility as you have more expertise here.

With regards to the headline, I would prefer for now not to make any significant visual changes, so maybe a first step would be to see how much we can improve accessibility at the HTML level without greatly changing the visual appearance. It might be possible with some styling to make an html heading work without too much wasted space, I can help with exploring this.

On a somewhat related note, I'm interested in eventually creating a replacement for the graph that would be accessible to blind people, that provides some summarizing statistics & information on the selected countries. Would like to prioritize that soon and think more about the best way to implement this. That's a separate issue but just wanted to let you know where I hope to go with this, in case it's relevant for decisions here.

**Comment by rpkoller** [comment]
No rush and worries :)  problem with the aside is that it is destined for related information like where the data is from or from what sources. so if you would put and aside and wrap the links to the johns hopkins university as data source or links to additional documentation to read up that would be appropriate but we have controls to directly alter the other elements content and its display. asked on stackoverflow two weeks ago. easiest is to read up the answer which is profound: https://stackoverflow.com/questions/61442767/what-is-the-proper-semantic-markup-for-a-statistical-graph-with-view-controls-al he is explaining the motivations behind the different steps. 

well the headline and using the h2 would be a reasonable step. i just havent performed any positioning work in css. meaning out of the box it takes too much space. from a visual perspective i consider it reasonable even necessary to move the most important and descriptive heading from a black color to red and especially to mark it that way semantically with an h2. well i just give it a try if i can accomplish a cleaner sticking more to the previous positioning version. 

and about the related notice. great! that is something i suggested above: 

> -generate such a descriptive text automatically based for each country selected.
> 
> describe todays current state with the number of overall infected persons and the number of infected persons last week and the number of deaths to date
> and if possible either state currently the growth is exponential or the growth is non exponential
> and slowing down... unsure if it is possible to calculate it in an either/or manner.
> there are many possibilities. but you could basically create snippets of text templates describing those different curves which are basically following all the same patterns. so if you could calculate and classify a curve with exponential growth or slowed down to non exponential growth and so on it would be possible. but that is just a future idea. ;)

Is there a way to calculate the following statement based on this: 

![EW4c7gEWkAA1KgX](https://user-images.githubusercontent.com/1665422/81800320-f58f9880-9512-11ea-8571-cbf011d00d07.jpeg)

spain: decreasing growth
NZ: disease transmission is under control
hong kong: exponential growth, decreasing, exponential growth, decreasing, exponential growth, disease transmission is under control. 

meaning that you segment and get the gist out of each curve and generate descriptive texts about the selected curves. something like 

New Zeeland: Has 1497 confirmed cases until the 11.5.2020. The rolling average of weekly confirmed cases within the last 7 days is 9. The transmission of covid 19 is contained and under control for 38 days 

-> the days since the curve went straight down. 

Germany: Has 173171 confirmed cases until  the 11.5.2020. The rolling average of weekly confirmed cases within the last 7 days is 6164. covid 19 has a decreasing growth for 44 days but is not stabilized yet.

and in case of something like singapore or hong kong we have to decide 
![Bildschirmfoto 2020-05-13 um 12 24 35](https://user-images.githubusercontent.com/1665422/81801472-ca0dad80-9514-11ea-9686-e79f49f3ae8c.png)
![Bildschirmfoto 2020-05-13 um 12 24 48](https://user-images.githubusercontent.com/1665422/81801477-cc700780-9514-11ea-8e7e-6e8c5b17e787.png)

how to describe. meaning segment the graphs and say first exponential then decreasing and so forth or if you generally the the current state at the end. but add before there were switched between growth types. 

each way and in general as described it would be possible in case you are able to calculate and classify the latest current growth type to create text snippet that could be pieced together accordingly.

**Comment by rpkoller** [comment]
and i will also investigate and test a little in regards of outlines focus states and so forth. that wasn't included in the first step/revision in regards of a11y neither
